File: src/competition/SeasonCalendar.cs
using ElevenLegends.Data.Enums;
using ElevenLegends.Data.Models;

namespace ElevenLegends.Competition;

/// <summary>
/// Represents a single day in the season calendar.
/// </summary>
public sealed record SeasonDay
{
    public required int Day { get; init; }
    public required DayType Type { get; init; }
    public IReadOnlyList<MatchFixture> Fixtures { get; init; } = [];
}

/// <summary>
/// Builds the full season calendar: national knockouts → mundial.
/// Structure: Training → Quarters → Training → Semis → Training → Final →
///            Training → Mundial Semis → Training → Mundial Final
/// </summary>
public static class SeasonCalendar
{
    /// <summary>
    /// Generates the complete season schedule. Match fixtures are filled in
    /// progressively as bracket rounds advance.
    /// Now includes a transfer window between nationals and mundial.
    /// </summary>
    public static List<SeasonDay> BuildTemplate()
    {
        var days = new List<SeasonDay>();
        int day = 1;

        // Phase 1: National knockouts
        days.AddRange(TrainingBlock(ref day, 3));
        days.Add(new SeasonDay { Day = day++, Type = DayType.MatchDay });     // Quarterfinals

        days.AddRange(TrainingBlock(ref day, 2));
        days.Add(new SeasonDay { Day = day++, Type = DayType.MatchDay });     // Semifinals

        days.AddRange(TrainingBlock(ref day, 2));
        days.Add(new SeasonDay { Day = day++, Type = DayType.MatchDay });     // Finals

        // Transfer window between nationals and mundial
        days.Add(new SeasonDay { Day = day++, Type = DayType.Rest });         // Rest before window
        for (int i = 0; i < 5; i++)
            days.Add(new SeasonDay { Day = day++, Type = DayType.TransferWindow });
        days.Add(new SeasonDay { Day = day++, Type = DayType.Rest });         // Rest after window

        // Phase 2: Mundial
        days.AddRange(TrainingBlock(ref day, 3));
        days.Add(new SeasonDay { Day = day++, Type = DayType.MundialMatchDay }); // Mundial Semis

        days.AddRange(TrainingBlock(ref day, 2));
        days.Add(new SeasonDay { Day = day++, Type = DayType.MundialMatchDay }); // Mundial Final

        return days;
    }

    private static List<SeasonDay> TrainingBlock(ref int day, int count)
    {
        var block = new List<SeasonDay>();
        for (int i = 0; i < count; i++)
        {
            block.Add(new SeasonDay { Day = day++, Type = DayType.Training });
        }
        return block;
    }
}

File: engine/competition/season-calendar.ts
import { DayType } from "@/engine/enums";
import type { MatchFixture } from "@/engine/models";

/**
 * Represents a single day in the season calendar.
 */
export interface SeasonDay {
  day: number;
  type: DayType;
  fixtures: readonly MatchFixture[];
}

/**
 * Builds the full season calendar: national knockouts -> mundial.
 * Structure: Training -> Quarters -> Training -> Semis -> Training -> Final ->
 *            Rest -> Transfer Window (5 days) -> Rest ->
 *            Training -> Mundial Semis -> Training -> Mundial Final
 */
export function buildTemplate(): SeasonDay[] {
  const days: SeasonDay[] = [];
  let day = 1;

  // Phase 1: National knockouts
  pushTrainingBlock(days, day, 3);
  day += 3;
  days.push({ day: day++, type: DayType.MatchDay, fixtures: [] }); // Quarterfinals

  pushTrainingBlock(days, day, 2);
  day += 2;
  days.push({ day: day++, type: DayType.MatchDay, fixtures: [] }); // Semifinals

  pushTrainingBlock(days, day, 2);
  day += 2;
  days.push({ day: day++, type: DayType.MatchDay, fixtures: [] }); // Finals

  // Transfer window between nationals and mundial
  days.push({ day: day++, type: DayType.Rest, fixtures: [] }); // Rest before window
  for (let i = 0; i < 5; i++) {
    days.push({ day: day++, type: DayType.TransferWindow, fixtures: [] });
  }
  days.push({ day: day++, type: DayType.Rest, fixtures: [] }); // Rest after window

  // Phase 2: Mundial
  pushTrainingBlock(days, day, 3);
  day += 3;
  days.push({ day: day++, type: DayType.MundialMatchDay, fixtures: [] }); // Mundial Semis

  pushTrainingBlock(days, day, 2);
  day += 2;
  days.push({ day: day++, type: DayType.MundialMatchDay, fixtures: [] }); // Mundial Final

  return days;
}

function pushTrainingBlock(
  days: SeasonDay[],
  startDay: number,
  count: number,
): void {
  for (let i = 0; i < count; i++) {
    days.push({ day: startDay + i, type: DayType.Training, fixtures: [] });
  }
}
